### PR TITLE
Change slice_hash_table to use linear probing and add unit test.

### PR DESCRIPTION
**Pull request by markdroth**
This fixes a bug in the old quadratic probing implementation that sometimes caused a crash when constructing a hash table.

While I was at it, I also cleaned up the API a bit.  In particular, I changed `grpc_slice_hash_table_create()` to take ownership of keys and values, which eliminated the need for the `copy_value()` method in the vtable.  And I also eliminated the unnecessary support for specifying the vtable on a per-key basis; there is now just one `destroy_value()` function, which is passed to `grpc_slice_hash_table_create()`.

In addition, I added some missing rules to `test/core/slice/BUILD` and fixed one existing rule that was pointing to the wrong file.

### CODE REVIEW

File: src/core/lib/slice/slice_hash_table.c [hunk 1]
```diff
@@ -118,8 +109,15 @@ void grpc_slice_hash_table_unref(grpc_exec_ctx* exec_ctx,
...
                                 const grpc_slice key) {
-  const size_t idx =
-      grpc_slice_hash_table_find_index(table, key, false /* find_empty */);
-  if (idx == table->size) return NULL;  // Not found.
-  return table->entries[idx].value;
+  const size_t hash = grpc_slice_hash(key);
+  // We cap the number of probes at the max number recorded when
+  // populating the table.
+  for (size_t offset = 0; offset <= table->max_num_probes; ++offset) {
+    const size_t idx = (hash + offset) % table->size;
```
Comment (dgquintas) [inline src/core/lib/slice/slice_hash_table.c]: the index calculation logic could be extracted into a small static function. It's repeated here and in the _add function.

File: test/core/slice/slice_hash_table_test.c [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    {"key_90", "value_90"},
+    {"key_91", "value_91"},
+    {"key_92", "value_92"},
+    {"key_93", "value_93"},
+    {"key_94", "value_94"},
+    {"key_95", "value_95"},
+    {"key_96", "value_96"},
+    {"key_97", "value_97"},
+    {"key_98", "value_98"},
+    {"key_99", "value_99"},
```
Comment (dgquintas) [inline test/core/slice/slice_hash_table_test.c]: I hope you didn't actually typed all these :)

File: src/core/lib/slice/slice_hash_table.c (hunk 1 above)
Comment (markdroth) [inline src/core/lib/slice/slice_hash_table.c]: I thought about that, but I don't think it's really worth doing.  The logic is fairly simple now, and we have tests to make sure it works, so I'm not worried about the two functions getting out of sync.

**Comment by markdroth** [comment]
Known issues: #10794 #10576

**Pull request merged**